<a href="https://colab.research.google.com/github/HydrOpOtesJA99/A.I-learning/blob/main/14_%EB%8D%B0%EC%9D%B4%ED%84%B0_%ED%83%80%EC%9E%85_%EC%A1%B0%EC%A0%95.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 14_데이터_타입_조정

> https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris2.csv

1. 변수 타입 확인 및 변경
  * 변수(칼럼) 타입확인: 데이터 dtypes
  * 변수를 범주형으로 변경:
    * 데이터['칼럼명'].astype('category')
  * 변수를 수치형으로 변경
    * 데이터['칼럼명'].astype('float') # 실수형
    * 데이터['칼럼명'].astype('int') # 정수형
2. NA 값의 처리
  * NA 갯수 체크: 데이터.isna().sum()
  * NA 값 채우기: 데이터['칼럼명'].fillna(특정숫자)



# 1. 변수 타입 확인 및 변경

In [1]:
import pandas as pd

In [4]:
파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris2.csv'
아이리스 = pd.read_csv(파일경로)
아이리스.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종
0   5.1  3.5    1.4   0.2   0
1   4.9  3.0    1.4   0.2   0
2   4.7  3.2    1.3   0.2   0
3   4.6  3.1    1.5   0.2   0
4   5.0  3.6    1.4   0.2   0



* 품종 data가 virginica, setosa 같이 명목형 데이터가 
* setosa=0, virginica=1, versicolor=2 와 같은 숫자형 데이터로 바뀐상태 
* but still 품종의 종류를 표시하는 **범주형** 데이터인 상태
* 따라서 실수형 데이터로 바꾸는 원핫인코딩 필요



## 1) 원핫인코딩

In [5]:
# 원핫인코딩

인코딩 = pd.get_dummies(아이리스)
인코딩.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종
0   5.1  3.5    1.4   0.2   0
1   4.9  3.0    1.4   0.2   0
2   4.7  3.2    1.3   0.2   0
3   4.6  3.1    1.5   0.2   0
4   5.0  3.6    1.4   0.2   0



*   but 원핫인코딩이 되지 않는다 왜냐하면 실제 data값은 실수로 적혀있기 때문이다



## 2) 변수의 타입확인

In [6]:
print(아이리스.dtypes)

꽃잎길이     float64
꽃잎폭      float64
꽃받침길이    float64
꽃받침폭     float64
품종         int64
dtype: object




*   float: 실수형, int: 정수형
*   object 또는 category의 경우에만 원핫인코딩가능



## 3) 변수를 범주형으로 변경

In [9]:
# 품종 타입을 범주형으로 바꾸어 준다.

아이리스['품종'] = 아이리스['품종'].astype('category')

print(아이리스.dtypes)

꽃잎길이      float64
꽃잎폭       float64
꽃받침길이     float64
꽃받침폭      float64
품종       category
dtype: object


In [11]:
# 원핫인코딩

인코딩 = pd.get_dummies(아이리스)
인코딩.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_0  품종_1  품종_2
0   5.1  3.5    1.4   0.2     1     0     0
1   4.9  3.0    1.4   0.2     1     0     0
2   4.7  3.2    1.3   0.2     1     0     0
3   4.6  3.1    1.5   0.2     1     0     0
4   5.0  3.6    1.4   0.2     1     0     0



*   원핫인코딩이 성공적으로 마무리



# 2. NA 값의 처리

## 1) NA 갯수 체크

In [12]:
# NA 값을 체크

아이리스.isna().sum()

꽃잎길이     0
꽃잎폭      1
꽃받침길이    0
꽃받침폭     0
품종       0
dtype: int64

In [13]:
아이리스.tail()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭 품종
145   6.7  3.0    5.2   2.3  2
146   6.3  2.5    5.0   1.9  2
147   6.5  3.0    5.2   2.0  2
148   6.2  3.4    5.4   2.3  2
149   5.9  NaN    5.1   1.8  2


*   149번째 행에 NaN(NA)값 확인
*   위 데이터를 모델에 그대로 넣으면 error 발생하므로 NA값을 바꾸어 줘야함



## 2) NA 값 채우기

In [14]:
# NA값에 꽃잎폭의 평균값을 넣어준다
mean = 아이리스['꽃잎폭'].mean()
아이리스['꽃잎폭'] = 아이리스['꽃잎폭'].fillna(mean) # NA값을 mean으로 채워라
아이리스.tail()

꽃잎길이       꽃잎폭  꽃받침길이  꽃받침폭 품종
145   6.7  3.000000    5.2   2.3  2
146   6.3  2.500000    5.0   1.9  2
147   6.5  3.000000    5.2   2.0  2
148   6.2  3.400000    5.4   2.3  2
149   5.9  3.054362    5.1   1.8  2

*   149번째 행에 NaN(NA)값을 '꽃잎폭'의 평균값으로 바꿈